app/services/fees/projection_service.rb

```diff
@@ -0,0 +1,173 @@
+# frozen_string_literal: true
+
+module Fees
+  class ProjectionService < ::BaseService
+    Result = BaseResult[:projected_amount_cents, :projected_units, :projected_pricing_unit_amount_cents]
+
+    def initialize(fees:)
+      @fees = fees
+      super(nil)
+    end
+
+    def call
+      result = Result.new
+
+      if charge&.billable_metric&.recurring?
+        current_amount_cents = fees.sum(&:amount_cents)
+        current_units = fees.sum { |f| BigDecimal(f.units) }
+        current_pricing_unit_amount_cents = fees.sum { |f| f.pricing_unit_usage&.amount_cents || 0 }
+        current_pricing_unit_amount_cents = nil if current_pricing_unit_amount_cents.zero?
+
+        result.projected_amount_cents = current_amount_cents
+        result.projected_units = current_units
+        result.projected_pricing_unit_amount_cents = current_pricing_unit_amount_cents
+        return result
+      end
+
+      if fees.blank? || !(period_ratio > 0 && period_ratio < 1)
+        result.projected_amount_cents = BigDecimal("0")
+        result.projected_units = BigDecimal("0")
+        result.projected_pricing_unit_amount_cents = BigDecimal("0")
+        return result
+      end
+
+      aggregation_result = run_aggregation
+      return result.fail_with_error!(aggregation_result.error) unless aggregation_result.success?
+
+      charge_model_result = Charges::ChargeModelFactory.new_instance(
+        charge: charge,
+        aggregation_result: aggregation_result,
+        properties: properties_for_charge_model,
+        period_ratio: period_ratio
+      ).apply
+
+      return result.fail_with_error!(charge_model_result.error) unless charge_model_result.success?
+
+      if charge_model_result.try(:grouped_results)
+        target_group_result = charge_model_result.grouped_results.find do |group_result|
+          group_result.grouped_by == first_fee.grouped_by
+        end
+        charge_model_result = target_group_result if target_group_result
+      end
+
+      result.projected_amount_cents = calculate_projected_amount_cents(charge_model_result)
+      result.projected_units = charge_model_result.projected_units
+      result.projected_pricing_unit_amount_cents = calculate_projected_pricing_unit_amount_cents(charge_model_result)
+      result
+    end
+
+    private
+
+    attr_reader :fees
+
+    def first_fee
```

Comment:
can we use it in the initialize method? so you can use the attr_reader, same for all the values that are fixed from the first_fee since it can't change after the service initialization